src/Spanner/ValueMapper.php

```diff
@@ -287,11 +289,14 @@ private function decodeValue($value, array $type)
      * Create a spanner parameter type value object from a PHP value type.
      *
      * @param mixed $value The PHP value
-     * @param int $givenType
-     * @param int $arrayType
+     * @param int $givenType If set, this type will be used in place of an inferred type.
+     * @param int $arrayType Defines the type of array elements.
+     * @param bool $allowMixedArrayType If true, array values may be of mixed type.
+     *        This is useful when reading against complex keys containing multiple
+     *        elements of differing types.
      * @return array The Value type
      */
-    private function paramType($value, $givenType = null, $arrayType = null)
+    private function paramType($value, $givenType = null, $arrayType = null, $allowMixedArrayType = false)
```

Comment:
Complex keys (i.e. keys containing >1 elements) are passed through this method before being sent to the API. Since the restrictions on single-type arrays don't apply to key elements, this additional argument allows that restriction to be bypassed when a keyset is being constructed.